File: jsAlgoDataStructuresMC/dataStructures/binaryHeap/priorityQueue.js
class Node{
    constructor(val, priority){
        this.priority = priority
        this.val = val
    }
}

class PriorityQueue{
    constructor(){
        this.queue = [];
    }
    swap(index1, index2){
        [this.queue[index1], this.queue[index2]] = [this.queue[index2], this.queue[index1]]
    }
    insert(val, priority){
        this.queue.push(new Node(val, priority))
        let child = this.queue.length - 1
        while(true){
            let parent = Math.floor((child - 1)/2)
            if(!this.queue[parent] || this.queue[child].priority < this.queue[parent].priority) return this.queue
            this.swap(child, parent)
            child = parent
        }
    }
    extractMax(){
        this.swap(0, this.queue.length - 1);
        if(this.queue.length === 1) return this.queue.pop();
        let popped = this.queue.pop();
        let parent = 0;
        while(true){
            let left = 2 * parent + 1;
            let right = 2 * parent + 2;
            let leftElement = this.queue[left] ? this.queue[left].priority : -Infinity;
            let rightElement = this.queue[right] ? this.queue[right].priority : -Infinity;
            let parentElement = this.queue[parent].priority;
            if(parentElement > leftElement && parentElement > rightElement) return popped
            if(leftElement >= rightElement){
                this.swap(parent, left)
                parent = left;
            }else if(rightElement > leftElement){
                this.swap(parent, right)
                parent = right;
            }
        }
    }
}

let pq = new PriorityQueue();
pq.insert(10, 0)
pq.insert(50, 1)
pq.insert(25, 10)
pq.insert(100, 5)
// console.log(pq)
console.log(pq.extractMax())
console.log(pq)
console.log(pq.extractMax())
console.log(pq)
console.log(pq.extractMax())
console.log(pq)
console.log(pq.extractMax())
console.log(pq)

File: jsAlgoDataStructuresMC/dataStructures/binaryHeap/priorityQueue.py
import math

class Node:
    def __init__(self, priority, val):
        self.priority = priority
        self.val = val
    def __str__(self):
        return f'Node [priority:{self.priority}, val:{self.val}]'

class PriorityQue:
    def __init__(self):
        self.queue = []
    
    def swap(self, index_1, index_2):
        self.queue[index_1], self.queue[index_2] = self.queue[index_2], self.queue[index_1]

    def insert(self, priority, val):
        self.queue.append(Node(priority, val))
        child = len(self.queue) - 1
        while True:
            parent = math.floor((child - 1) / 2) if math.floor((child - 1) / 2) > 0 else 0
            if self.queue[child].priority > self.queue[parent].priority:
                self.swap(child, parent)
                child = parent
            else:
                return self.queue
    
    #* Helps us print the queue a little neater
    def printQueue(self):
        for i, node in enumerate(self.queue): print(f'Position {i} - {node}')

pq = PriorityQue()
pq.insert(0, 10)
pq.insert(1, 20)
pq.insert(1, 20)
pq.insert(1, 20)
pq.insert(10, 20)
pq.printQueue()